src/handlers/shared/structured-metadata.ts

```diff
@@ -0,0 +1,35 @@
+import { LogLevel } from "../../adapters/supabase/pretty-logs";
+
+type Metadata<T extends object> = T & {
+  revision?: string;
+  logMessage?: {
+    type?: LogLevel;
+    message?: string;
+  };
+  [key: string]: unknown;
+};
+
+function createStructuredMetadata<T extends object>(className: string, metadata: Metadata<T>) {
+  const jsonPretty = JSON.stringify(metadata, null, 2);
+  const stackLine = new Error().stack?.split("\n")[2] ?? "";
+  const caller = stackLine.match(/at (\S+)/)?.[1] ?? "";
+  const ubiquityMetadataHeader = `<!-- Ubiquity - ${className} - ${caller} - ${metadata.revision}`;
+
+  let metadataSerialized: string;
+  const metadataSerializedVisible = ["```json", jsonPretty, "```"].join("\n");
+  const metadataSerializedHidden = [ubiquityMetadataHeader, jsonPretty, "-->"].join("\n");
+
+  if (metadata.logMessage?.type === LogLevel.FATAL) {
+    // if the log message is fatal, then we want to show the metadata
+    metadataSerialized = [metadataSerializedVisible, metadataSerializedHidden].join("\n");
+  } else {
+    // otherwise we want to hide it
+    metadataSerialized = metadataSerializedHidden;
+  }
+
+  return metadataSerialized;
+}
+
+export default {
+  create: createStructuredMetadata,
+};
```

Comment:
I wonder if we should proceed with the metadata system. 

It seems  really nice to have but we need to make it way more convenient for developers to work with (via our SDK) for this to catch on and be useful.

Reply:
I agree it should be integrated to the SDK to be more beneficial

It would be very effective as an error wrapper/fallback of sorts. Called in catch blocks etc and before the end of plugin execution it handles posting/storage/notifying making using it a breeze

With the above actually you could hoover up metadata throughout the whole plugin flow which could be v useful (logs would do this but it would save hunting for them)

Reply:
Isn't this solved, at least partially, by the updates you did to `ubiquibot-logger` @Keyrxng ?

Reply:
from a non-comment posting standpoint yes more or less this logic exists in `Logs.ts` and is returned under `metadata` in the `LogReturn` object from any Log call. It includes manipulating the stack etc but it doesn't include logic for formatting to hide content in a comment it's just a diff at the minute.

But you make a good point in that it could also just be handled by the logger which is expected to be used in all plugins